### ISSUE TITLE
RuntimeError in writing threads in asyncwrite doesn't quit job immediately

### ISSUE THREAD

**Issue by arkal** [BD2KGenomics/toil#482]
If a RuntimeError is raised by a thread, the program should quit immediately instead of waiting for remaining data to be written to the fileStore.

If the reasoning behind it NOT doing so is that other non-error data is written to filestore and picked up on restart, then there (might be) a bug in the code because --restart did not detect the files written in the previous attempt and rewrote all files to the filestore. However, in the previous run 4 threads returned a RuntimeError and I cannot say what file they each were working on.

**Comment by arkal** [comment]
This is a ```watch tree-fh /path/to/jobstore``` and the two 4.5G and the 25G objects are my fastqs and my index.tar.gz respectively. The 4.0 and 4.1G objects are the rewritten fastqs in the --restart ed job and the 5.0 is the rewritten index
![tree structure of filestore]
(http://i58.tinypic.com/ejbvrs.png)

**Comment by benedictpaten** [comment]
Hi Arjun,

I'm unclear on this. Are you saying it's a bug? Yes, it's true that the
asynchronous writes won't quite immediately when another thread terminates,
but that doesn't/shouldn't matter, aside from being a bit annoying having
to wait. There is no reliable "thread kill" in python (apart from
sigkill!), so getting a asynchronous file transfer to exit immediately is
non-trivial.

Can I see the stack trace?

**Comment by arkal** [comment]
Ok cool. I marked it as both bug and enhancement depending on which paragraph above was relevant. 

I could see this being annoying to wait if the bamfile is 300G (especially if we're paying for compute time and marked it for enhancement.  It would be a bug if you were doing the --reset thing i spoke about. 

The stack trace will have something like this pop up
```
Exception in thread Thread-12:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/local/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/site-packages/toil/job.py", line 600, in asyncUpdate
    raise RuntimeError("The termination flag is set, exiting before update")
RuntimeError: The termination flag is set, exiting before update
```
Then you have to wait for all threads to join before the job really ends with an error.

**Comment by hannes-ucsc** [comment]
We should start factoring out flow control functionality like killable threads. Then we can get fancy without spewing that stuff all over the code. A pseudo-killable Thread could be implemented neatly in a custom Thread subclass. To use it, you would need to sprinkle your code, especially loop bodies with call to threadKilled(). The loop that copies to/fro the cache should check every megabyte or so. 

Let me know if you want me to prototype one.

**Comment by benedictpaten** [comment]
Hi Arjun, I still don't think the restart case is a bug, it just wastes a bunch of time - but it does not lead to a crash. It is absolutely intended that if a job fails all files written by it are essentially ignored, even if they were needlessly written after a crash is detected. The only aspect I consider a "bug" here is that the partially written files are not cleaned up. It would be good to have a thread cleaning the file store in the case of partial writes.  

To Hanne's point - sure, if prototyping this would be good. I'm removing the "bug" flag here and assigning to Hannes, but putting on the back log. Is that okay?

**Comment by arkal** [comment]
Yes, Thanks!

**Comment by hannes-ucsc** [comment]
Sounds good.

**Comment by hannes-ucsc** [comment]
Should be resolved together with #481 and #474.

**Comment by arkal** [comment]
Outdated, possibly naive comment. Closing.
